### ISSUE TITLE
Behavior of next_series_episodes with release history of season packs vs episodes

### ISSUE THREAD

**Issue by tubedogg** [Flexget/Flexget#1782]
Continued discussion from #1776 since the immediate issue there was resolved.

### Expected behaviour:
Given a release history of season entities S01 and S02 for My Show in the series database with an identifier type of ep, ``next_series_episodes`` will emit S03E01 as the next episode to obtain, and ``next_series_seasons`` will emit S03 as the next season to obtain.

### Actual behaviour:
Both plugins emit nothing unless ``from_start`` option is set or a begin episode has been specified. This is unintuitive and works counter to how ``next_series_episodes`` operates when episode releases are present in the series database. (See further discussion at the bottom.) At the very least, the error message stating the show has no history is incorrect.

### Steps to reproduce:
- Step 1:
```
→ flexget inject "My Show S01 720p hdtv h264 dd5.1" --tasks learn_eps --learn --force --disable-tracking
→ flexget inject "My Show S02 720p hdtv h264 dd5.1" --tasks learn_eps --learn --force --disable-tracking
```
- Step 2: Run test_nse task (I used fake BTN info here because the task stops before it ever gets there).

#### Config:
```
tasks:
  learn_eps:
    all_series:
      season_packs: always
    manual: yes

  test_nse:
    discover:
      what:
        - next_series_episodes: yes
      from:
        - btn: MyAPIkey
    series:
      - My Show:
           identified_by: ep
```
  
#### Log:
```
2017-04-06 18:11 DEBUG    series        test_nse no downloaded episodes found for series 'My Show', season: None, downloaded: True
2017-04-06 18:11 DEBUG    series        test_nse no downloaded episodes found for series 'My Show', season: 3, downloaded: True
2017-04-06 18:11 DEBUG    series        test_nse no downloaded season packs found for series 'My Show', season: 3, downloaded: True
2017-04-06 18:11 VERBOSE  next_series_episodes test_nse Series `My Show` has no history. Set begin option, or use CLI `series begin` subcommand to set first episode to emit
2017-04-06 18:11 WARNING  discover      test_nse Input next_series_episodes did not return anything
```

### Additional information:

- Flexget Version: 2.10.29
- Python Version: 2.7.13
- Installation method: pip
- OS and version: macOS El Capitan 10.11.6
- Link to crash log: n/a

### Further discussion
Scenarios:
1a. I insert My Show S01E01 into the database. I don't set any begin options or anything like that.
1b. I run a discover task with next_series_episodes on it. next_series_episodes emits S01E02 as the next expected episode and discover starts looking for that.
1c. As a result of using estimators, discover figures out that each season is only one episode, and then starts looking for S02E01.

2a. I insert My Show S01 into the database. Again, no begin options or anything.
2b. I run a discover task with next_series_episodes on it. next_series_episodes wants me to set a begin episode.

There's inconsistency between these two nearly-identical scenarios. In both cases, the database knows the identifier style. (I even added it to the configure_series settings in the task to be doubly sure.) In both cases, a season has ended and it should be looking for the next one - scenario 1 doesn't know that and outputs S01E02 and uses estimators to find out that's not real and to now check for S02E01 ... but scenario 2 doesn't even output S02E01 as a possible starting point, it just gives up. But scenario 2 knows I have completed S01, so why _wouldn't_ it output S02E01 and let the estimators sort out if that's a real episode?

It seems like the assumption is that if there are episodes downloaded in the latest known season, to continue downloading there, but if all known seasons are completed, to not try to start a new season (unless ``from_start`` is set). Is that correct?

Contrast this with the normal ``series`` plugin, which given scenario 2 above accepts S02E01 as the next episode.

After I read the code, I added ``from_start: yes``, and it immediately worked ... but I'm confused why this is necessary to get the two scenarios to work in a consistent manner. Looking at next_series_seasons code, it seems like if you only ever download season packs, next_series_seasons just flat-out wouldn't work (again without using from_start).

It's not that I'm necessarily averse to adding from_start to the config, but the option name wouldn't jump out at me as something that I should add when (as in my original case with Review) I don't actually want to get downloads from the start. It also seems very unintuitive that episodes and season packs are treated so differently - episode history works by itself, but season pack history is meaningless without additional options set.

Maybe I'm completely off-base here and there are other reasons for this being designed as it is. I've been wrong before. :)

**Comment by liiight** [comment]
I agree that this is a new behavior pattern but I'm not sure that it is handled incorrectly now. 
The scenario you presented is not akin to a season with a downloaded episode, which in turn tell `next_series_*` to look in current and next season on rerun, it's more analogous to a series with no history at all. 

Imagine a scenario where s03 and s06 were accepted somehow. What would you imagine the next emit season/ep should be? Without setting begin episode, there is no way to decide that. 

That being said I'm obviously not fixed on anything as this is brand new grounds. It may very well be that some base decisions regarding these plugins need to change. @gazpachoking @paranoidi I'd love to get your feedback on this.

**Comment by tubedogg** [comment]
Thanks for the feedback. :)

If ``backfill`` is not set, it should emit S07 or S07E01 (just like ``series`` would accept either of those as being the next thing after S06 and would ignore the gap that occurred _before_ the most recently-downloaded entity). If you have S03E01 and S06E01 in history, it would emit S06E02 without any issue, so the disconnect is what changes just because you're dealing with a different size entity (seasons)?

Another way of thinking about it is episodes exist within a definable container: a season. Seasons also exist within a definable container: the show itself. From this perspective both episodes and seasons are logical, consecutive units within their respective containers. Treating them differently doesn't make any sense, at least to me.

Regarding ``begin``, to me that is specifically for cases where you want to skip around from the current history. If I set ``begin S07E01`` and it downloads S07 for me, what happens when season 8 rolls around? Currently I would have to go back and set ``begin S08E01`` for ``next_series_*`` to do anything. This is why treating seasons differently doesn't work long-term. 

(And should we stick with how it currently works, ``from_start`` should be renamed [``continue_series``?] as it doesn't actually mean that in the context of this discussion.)

Anyway I'll shut up now and let others weigh in. :D

**Comment by jawilson** [comment]
I came across this recently and I definitely agree with @tubedogg that the behavior he describes makes more sense.

**Comment by liiight** [comment]
Yea, I agree. I'm pretty busy lately but I'll try to get to it ASAP.

**Comment by tubedogg** [comment]
I think it's just a matter of adding a couple lines. I'm gonna take a stab at it and will submit a PR.

**Comment by tubedogg** [comment]
Closed in #1797.
